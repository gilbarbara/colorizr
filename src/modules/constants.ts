import { ColorKeysTuple, ColorModelKey } from '~/types';

export const COLOR_KEYS: Record<ColorModelKey, ColorKeysTuple> = {
  hsl: ['h', 's', 'l'],
  oklab: ['l', 'a', 'b'],
  oklch: ['l', 'c', 'h'],
  rgb: ['r', 'g', 'b'],
};
export const COLOR_MODELS: ColorModelKey[] = ['hsl', 'oklab', 'oklch', 'rgb'];
export const DEG2RAD = Math.PI / 180;
export const LAB_TO_LMS = {
  l: [0.3963377773761749, 0.2158037573099136],
  m: [-0.1055613458156586, -0.0638541728258133],
  s: [-0.0894841775298119, -1.2914855480194092],
};
export const LRGB_TO_LMS = {
  l: [0.4122214708, 0.5363325363, 0.0514459929],
  m: [0.2119034982, 0.6806995451, 0.1073969566],
  s: [0.0883024619, 0.2817188376, 0.6299787005],
};
export const LSM_TO_LAB = {
  l: [0.2104542553, 0.793617785, 0.0040720468],
  a: [1.9779984951, 2.428592205, 0.4505937099],
  b: [0.0259040371, 0.7827717662, 0.808675766],
};
export const LSM_TO_RGB = {
  r: [4.0767416360759583, -3.3077115392580629, 0.2309699031821043],
  g: [-1.2684379732850315, 2.6097573492876882, -0.341319376002657],
  b: [-0.0041960761386756, -0.7034186179359362, 1.7076146940746117],
};
export const PRECISION = 5;
export const RAD2DEG = 180 / Math.PI;

export const MESSAGES = {
  alpha: 'amount must be a number between 0 and 1',
  left: 'left is required and must be a string',
  right: 'right is required and must be a string',
  input: 'input is required',
  inputHex: 'input is required and must be a hex',
  inputNumber: 'input is required and must be a number',
  inputString: 'input is required and must be a string',
  invalid: 'invalid input',
  invalidCSS: 'invalid CSS string',
  options: 'invalid options',
  threshold: 'threshold must be a number between 0 and 255',
} as const;
