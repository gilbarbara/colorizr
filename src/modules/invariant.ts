export function invariant(condition: boolean, message: string): asserts condition {
  if (condition) {
    return;
  }

  /* istanbul ignore else */
  if (process.env.NODE_ENV !== 'production') {
    if (message === undefined) {
      throw new Error('invariant requires an error message argument');
    }
  }

  const error = !message
    ? new Error(
        'Minified exception occurred; use the non-minified dev environment for the full error message and additional helpful warnings.',
      )
    : new Error(message);

  error.name = 'colorizr';

  throw error;
}
