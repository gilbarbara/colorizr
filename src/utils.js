// @flow
/**
 * Colorizr Utils
 * @module Utils
 */
/**
 * Parse math string expressions.
 *
 * @function
 * @param {string} input
 *
 * @returns {number}
 */
export function expr(input: string): number {
  const chars = input.split('');
  const n = [];
  const op = [];

  let parsed;
  let index = 0;
  let last = true;

  n[index] = '';

  // Parse the string
  for (let c = 0; c < chars.length; c++) {
    if (isNaN(parseInt(chars[c], 10)) && chars[c] !== '.' && !last) {
      op[index] = chars[c];
      index++;
      n[index] = '';
      last = true;
    }
    else {
      n[index] += chars[c];
      last = false;
    }
  }

  // Calculate the expression
  parsed = parseFloat(n[0]);
  for (let o = 0; o < op.length; o++) {
    const num = parseFloat(n[o + 1]);

    switch (op[o]) {
      case '+':
        parsed += num;
        break;
      case '-':
        parsed -= num;
        break;
      case '*':
        parsed *= num;
        break;
      case '/':
        parsed /= num;
        break;
      default:
        break;
    }
  }

  return parsed;
}

/**
 * Check if the input is an object
 *
 * @param {*} input
 * @returns {boolean}
 */
export function isPlainObject(input: any): boolean {
  if (!input) return false;

  const { toString } = Object.prototype;
  const prototype = Object.getPrototypeOf(input);

  return toString.call(input) === '[object Object]' && (prototype === null || prototype === Object.getPrototypeOf({}));
}

export function isRequired(name: string) {
  throw new Error(`${name || 'parameter'} is required`);
}

/**
 * Creates an object composed of the picked source properties.
 *
 * @param {Object} input
 * @param {Array} options
 *
 * @returns {Object}
 */
export function pick(input: Object, options: Array<string>): Object {
  if (!Array.isArray(options)) {
    throw new Error('options must be an array');
  }

  return options
    .filter(d => typeof input[d] !== 'undefined')
    .reduce((acc, i) => {
      acc[i] = input[i];
      return acc;
    }, {});
}

/**
 * Round decimal numbers.
 *
 * @function
 * @param {number} input
 * @param {number} digits
 *
 * @returns {number}
 */
export function round(input: number, digits: number = 2): number {
  const factor = 10 ** digits;
  return Math.round(input * factor) / factor;
}

export function validateHex(input: string): boolean {
  return /(^#[0-9A-F]{6}$)|(^#[0-9A-F]{3}$)/i.test(input);
}
