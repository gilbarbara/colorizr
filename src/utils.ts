import { HSL, RGB, PlainObject, RGBArray } from './types';

export const HSLKeys = ['h', 's', 'l'];
export const RGBKeys = ['r', 'g', 'b'];

/**
 * Constrain value into the range
 */
export function constrain(input: number, amount: number, range: number[], sign: string): number {
  invariant(arguments.length < 4, 'All parameters are required');

  const [min, max] = range;
  let value = expr(input + sign + amount);

  if (value < min) {
    value = min;
  } else if (value > max) {
    value = max;
  }

  return Math.abs(value);
}

/**
 * Constrain an angle
 */
export function constrainDegrees(input: number, amount: number): number {
  invariant(!isNumber(input), 'input is required');

  let value = input + amount;

  if (value > 360) {
    value %= 360;
  }

  if (value < 0) {
    value += 360;
  }

  return Math.abs(value);
}

/**
 * CSS named colors
 */
export const cssColors = {
  aliceblue: '#f0f8ff',
  antiquewhite: '#faebd7',
  aqua: '#00ffff',
  aquamarine: '#7fffd4',
  azure: '#f0ffff',
  beige: '#f5f5dc',
  bisque: '#ffe4c4',
  black: '#000000',
  blanchedalmond: '#ffebcd',
  blue: '#0000ff',
  blueviolet: '#8a2be2',
  brown: '#a52a2a',
  burlywood: '#deb887',
  cadetblue: '#5f9ea0',
  chartreuse: '#7fff00',
  chocolate: '#d2691e',
  coral: '#ff7f50',
  cornflowerblue: '#6495ed',
  cornsilk: '#fff8dc',
  crimson: '#dc143c',
  cyan: '#00ffff',
  darkblue: '#00008b',
  darkcyan: '#008b8b',
  darkgoldenrod: '#b8860b',
  darkgray: '#a9a9a9',
  darkgrey: '#a9a9a9',
  darkgreen: '#006400',
  darkkhaki: '#bdb76b',
  darkmagenta: '#8b008b',
  darkolivegreen: '#556b2f',
  darkorange: '#ff8c00',
  darkorchid: '#9932cc',
  darkred: '#8b0000',
  darksalmon: '#e9967a',
  darkseagreen: '#8fbc8f',
  darkslateblue: '#483d8b',
  darkslategray: '#2f4f4f',
  darkslategrey: '#2f4f4f',
  darkturquoise: '#00ced1',
  darkviolet: '#9400d3',
  deeppink: '#ff1493',
  deepskyblue: '#00bfff',
  dimgray: '#696969',
  dimgrey: '#696969',
  dodgerblue: '#1e90ff',
  firebrick: '#b22222',
  floralwhite: '#fffaf0',
  forestgreen: '#228b22',
  fuchsia: '#ff00ff',
  gainsboro: '#dcdcdc',
  ghostwhite: '#f8f8ff',
  gold: '#ffd700',
  goldenrod: '#daa520',
  gray: '#808080',
  grey: '#808080',
  green: '#008000',
  greenyellow: '#adff2f',
  honeydew: '#f0fff0',
  hotpink: '#ff69b4',
  indianred: '#cd5c5c',
  indigo: '#4b0082',
  ivory: '#fffff0',
  khaki: '#f0e68c',
  lavender: '#e6e6fa',
  lavenderblush: '#fff0f5',
  lawngreen: '#7cfc00',
  lemonchiffon: '#fffacd',
  lightblue: '#add8e6',
  lightcoral: '#f08080',
  lightcyan: '#e0ffff',
  lightgoldenrodyellow: '#FAFAD2',
  lightgray: '#d3d3d3',
  lightgrey: '#d3d3d3',
  lightgreen: '#90ee90',
  lightpink: '#ffb6c1',
  lightsalmon: '#ffa07a',
  lightseagreen: '#20b2aa',
  lightskyblue: '#87cefa',
  lightslategray: '#778899',
  lightslategrey: '#778899',
  lightsteelblue: '#b0c4de',
  lightyellow: '#ffffe0',
  lime: '#00ff00',
  limegreen: '#32cd32',
  linen: '#faf0e6',
  magenta: '#ff00ff',
  maroon: '#800000',
  mediumaquamarine: '#66cdaa',
  mediumblue: '#0000cd',
  mediumorchid: '#ba55d3',
  mediumpurple: '#9370d8',
  mediumseagreen: '#3cb371',
  mediumslateblue: '#7b68ee',
  mediumspringgreen: '#00fa9a',
  mediumturquoise: '#48d1cc',
  mediumvioletred: '#c71585',
  midnightblue: '#191970',
  mintcream: '#f5fffa',
  mistyrose: '#ffe4e1',
  moccasin: '#ffe4b5',
  navajowhite: '#ffdead',
  navy: '#000080',
  oldlace: '#fdf5e6',
  olive: '#808000',
  olivedrab: '#6b8e23',
  orange: '#ffa500',
  orangered: '#ff4500',
  orchid: '#da70d6',
  palegoldenrod: '#eee8aa',
  palegreen: '#98fb98',
  paleturquoise: '#afeeee',
  palevioletred: '#d87093',
  papayawhip: '#ffefd5',
  peachpuff: '#ffdab9',
  peru: '#cd853f',
  pink: '#ffc0cb',
  plum: '#dda0dd',
  powderblue: '#b0e0e6',
  purple: '#800080',
  red: '#ff0000',
  rosybrown: '#bc8f8f',
  royalblue: '#4169e1',
  saddlebrown: '#8b4513',
  salmon: '#fa8072',
  sandybrown: '#f4a460',
  seagreen: '#2e8b57',
  seashell: '#fff5ee',
  sienna: '#a0522d',
  silver: '#c0c0c0',
  skyblue: '#87ceeb',
  slateblue: '#6a5acd',
  slategray: '#708090',
  slategrey: '#708090',
  snow: '#fffafa',
  springgreen: '#00ff7f',
  steelblue: '#4682b4',
  tan: '#d2b48c',
  teal: '#008080',
  thistle: '#d8bfd8',
  tomato: '#ff6347',
  turquoise: '#40e0d0',
  violet: '#ee82ee',
  wheat: '#f5deb3',
  white: '#ffffff',
  whitesmoke: '#f5f5f5',
  yellow: '#ffff00',
  yellowgreen: '#9acd32',
};

/**
 * Parse math string expressions
 */
export function expr(input: string): number {
  const chars = input.split('');
  const n: string[] = [];
  const op: string[] = [];

  let parsed;
  let index = 0;
  let last = true;

  n[index] = '';

  // Parse the string
  for (const char of chars) {
    if (Number.isNaN(parseInt(char, 10)) && char !== '.' && !last) {
      op[index] = char;
      index++;
      n[index] = '';
      last = true;
    } else {
      n[index] += char;
      last = false;
    }
  }

  // Calculate the expression
  parsed = parseFloat(n[0]);
  for (let o = 0; o < op.length; o++) {
    const num = parseFloat(n[o + 1]);

    switch (op[o]) {
      case '+':
        parsed += num;
        break;
      case '-':
        parsed -= num;
        break;
      case '*':
        parsed *= num;
        break;
      case '/':
        parsed /= num;
        break;
      default:
        break;
    }
  }

  return parsed;
}

export function invariant(
  condition: boolean,
  message: string,
  a?: any,
  b?: any,
  c?: any,
  d?: any,
  e?: any,
  f?: any,
): void {
  /* istanbul ignore else */
  if (process.env.NODE_ENV !== 'production') {
    if (message === undefined) {
      throw new Error('invariant requires an error message argument');
    }
  }

  if (condition) {
    let error;

    if (!message) {
      error = new Error(
        'Minified exception occurred; use the non-minified dev environment ' +
          'for the full error message and additional helpful warnings.',
      );
    } else {
      const args = [a, b, c, d, e, f];
      let argIndex = 0;
      error = new Error(message.replace(/%s/g, () => args[argIndex++]));
      error.name = 'colorizr';
    }

    throw error;
  }
}

/**
 * Check if an object contains HSL values
 */
export function isHSL(input: any): input is HSL {
  if (!isPlainObject(input)) {
    return false;
  }

  const entries = Object.entries(input);

  return (
    !!entries.length &&
    entries.every(
      ([key, value]) =>
        HSLKeys.indexOf(key) >= 0 && value >= 0 && value <= (key === 'h' ? 360 : 100),
    )
  );
}

/**
 * Check if the input is a number and not NaN
 */
export function isNumber(input: any): input is number {
  return typeof input === 'number' && !Number.isNaN(input);
}

/**
 * Check if the input is an object
 */
export function isPlainObject(input: any): input is PlainObject {
  if (!input) {
    return false;
  }

  const { toString } = Object.prototype;
  const prototype = Object.getPrototypeOf(input);

  return (
    toString.call(input) === '[object Object]' &&
    (prototype === null || prototype === Object.getPrototypeOf({}))
  );
}

/**
 * Check if an object contains RGB values.
 */
export function isRGB(input: any): input is RGB {
  if (!isPlainObject(input)) {
    return false;
  }

  const entries = Object.entries(input);

  return (
    !!entries.length &&
    entries.every(([key, value]) => RGBKeys.indexOf(key) >= 0 && value >= 0 && value <= 255)
  );
}

/**
 * Check if an array contains RGB values.
 */
export function isRGBArray(input: any): input is RGBArray {
  return Array.isArray(input) && input.length === 3 && input.every(d => d >= 0 && d <= 255);
}

/**
 * Check if the input is a string
 */
export function isString(input: any): input is string {
  return typeof input === 'string';
}

/**
 * Limit values per type.
 */
export function limit(input: number, type: string): number {
  invariant(!isNumber(input), 'Input is not a number');

  /* istanbul ignore else */
  if (RGBKeys.indexOf(type) >= 0) {
    return Math.max(Math.min(input, 255), 0);
  }

  if (['s', 'l'].indexOf(type) >= 0) {
    return Math.max(Math.min(input, 100), 0);
  }

  if (type === 'h') {
    return Math.max(Math.min(input, 360), 0);
  }

  throw new Error('Invalid type');
}

export const messages = {
  amount: 'amount must be a number',
  left: 'left is required and must be a string',
  right: 'right is required and must be a string',
  input: 'input is required',
  inputString: 'input is required and must be a string',
  invalid: 'invalid input',
  options: 'invalid options',
};

/**
 * Creates an object composed of the picked source properties.
 */
export function pick(input: PlainObject, options: string[]): PlainObject {
  if (!Array.isArray(options)) {
    throw new Error('options must be an array');
  }

  return options
    .filter(d => typeof input[d] !== 'undefined')
    .reduce((acc: PlainObject, d) => {
      acc[d] = input[d];
      return acc;
    }, {});
}

/**
 * Round decimal numbers.
 */
export function round(input: number, digits = 2): number {
  const factor = 10 ** digits;
  return Math.round(input * factor) / factor;
}
