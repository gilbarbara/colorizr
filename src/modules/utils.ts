import { COLOR_KEYS, COLOR_MODELS, MESSAGES, PRECISION } from '~/modules/constants';
import { invariant } from '~/modules/invariant';
import {
  isHSL,
  isLAB,
  isLCH,
  isNumber,
  isPlainObject,
  isRGB,
  isValidColorModel,
} from '~/modules/validators';

import {
  Alpha,
  ColorModel,
  ColorModelKey,
  ColorModelKeys,
  ConverterParameters,
  LAB,
  LCH,
  PlainObject,
} from '~/types';

export function addAlpha<T extends ColorModel>(input: any, alpha?: Alpha): T {
  invariant(isValidColorModel(input), MESSAGES.invalid);

  let value = alpha;

  if (!value) {
    return input;
  }

  /* c8 ignore next 3 */
  if (value > 1) {
    value /= 100;
  }

  if (value === 1) {
    return input;
  }

  return { ...input, alpha: value };
}

/**
 * Clamp a value between a min and max
 * @param value
 * @param [min=0] - The minimum value
 * @param [max=100] - The maximum value
 */
export function clamp(value: number, min = 0, max = 100) {
  return Math.min(Math.max(value, min), max);
}

/**
 * Constrain the degrees between 0 and 360
 */
export function constrainDegrees(input: number, amount: number): number {
  invariant(isNumber(input), MESSAGES.inputNumber);

  let value = input + amount;

  if (value > 360) {
    value %= 360;
  }

  if (value < 0) {
    value += 360;
  }

  return Math.abs(value);
}

/**
 * Limit values per type.
 */
export function limit<TModel extends Extract<ColorModelKey, 'hsl' | 'rgb'>>(
  input: number,
  model: TModel,
  key: ColorModelKeys<TModel>,
): number {
  invariant(isNumber(input), 'Input is not a number');
  invariant(COLOR_MODELS.includes(model), `Invalid model${model ? `: ${model}` : ''}`);
  invariant(COLOR_KEYS[model].includes(key), `Invalid key${key ? `: ${key}` : ''}`);

  switch (model) {
    case 'hsl': {
      invariant(COLOR_KEYS.hsl.includes(key), 'Invalid key');

      if (['s', 'l'].includes(key)) {
        return clamp(input);
      }

      return clamp(input, 0, 360);
    }
    case 'rgb': {
      invariant(COLOR_KEYS.rgb.includes(key), 'Invalid key');

      return clamp(input, 0, 255);
    }
    /* c8 ignore next 3 */
    default: {
      throw new Error('Invalid inputs');
    }
  }
}

/**
 * Parse the input parameters
 */
export function parseInput<T extends ColorModel>(
  input: ConverterParameters<T>,
  model: ColorModelKey,
): T {
  const keys = COLOR_KEYS[model];
  const validator = {
    hsl: isHSL,
    oklab: isLAB,
    oklch: isLCH,
    rgb: isRGB,
  };

  invariant(isPlainObject(input) || Array.isArray(input), MESSAGES.invalid);

  const value = Array.isArray(input)
    ? ({ [keys[0]]: input[0], [keys[1]]: input[1], [keys[2]]: input[2] } as unknown as T)
    : input;

  invariant(validator[model](value), `invalid ${model} color`);

  return value;
}

/**
 * Creates an object composed of the picked source properties.
 */
export function pick(input: PlainObject, options: string[]): PlainObject {
  if (!Array.isArray(options)) {
    throw new TypeError('options must be an array');
  }

  return options
    .filter(d => typeof input[d] !== 'undefined')
    .reduce((acc: PlainObject, d) => {
      acc[d] = input[d];

      return acc;
    }, {});
}

/**
 * Restrict the values to a certain number of digits.
 */
export function restrictValues<T extends LAB | LCH>(
  input: T,
  precision: number = PRECISION,
  forcePrecision = true,
): T {
  const output = new Map(Object.entries(input));

  for (const [key, value] of output.entries()) {
    output.set(key, round(value, precision, forcePrecision));
  }

  return Object.fromEntries(output) as T;
}

/**
 * Round decimal numbers.
 */
export function round(input: number, precision = 2, forcePrecision = true): number {
  if (!isNumber(input) || input === 0) {
    return 0;
  }

  if (forcePrecision) {
    const factor = 10 ** precision;

    return Math.round(input * factor) / factor;
  }

  const absInput = Math.abs(input);

  let digits = Math.abs(Math.ceil(Math.log(absInput) / Math.LN10));

  if (digits === 0) {
    digits = 2;
  } else if (digits > precision) {
    digits = precision;
  }

  let exponent = precision - (digits < 0 ? 0 : digits);

  if (exponent <= 1 && precision > 1) {
    exponent = 2;
  } else if (exponent > precision || exponent === 0) {
    exponent = precision;
  }

  const factor = 10 ** exponent;

  return Math.round(input * factor) / factor;
}
