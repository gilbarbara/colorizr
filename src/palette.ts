import { MESSAGES } from '~/modules/constants';
import { invariant } from '~/modules/invariant';
import { isHex, isNamedColor, isPlainObject, isString } from '~/modules/validators';

import convert from '~/convert';
import hex2hsl from '~/converters/hex2hsl';
import hsl2hex from '~/converters/hsl2hex';
import extractColorParts from '~/extract-color-parts';
import parseCSS from '~/parse-css';
import rotate from '~/rotate';
import { ColorType, HEX } from '~/types';

export interface PaletteOptions {
  format?: ColorType;
  lightness?: number;
  saturation?: number;
  /**
   * The number of colors to generate
   * @default 6
   */
  size?: number;
  type?: 'monochromatic';
}

export default function palette(input: string, options: PaletteOptions = {}): string[] {
  invariant(isString(input), MESSAGES.inputString);
  invariant(isPlainObject(options), MESSAGES.options);

  const { format, lightness, saturation, size = 6, type } = options;
  const hsl = parseCSS(input, 'hsl');
  const output: string[] = [];
  const colorFormat = isHex(input) || isNamedColor(input) ? 'hex' : extractColorParts(input).model;

  switch (type) {
    case 'monochromatic': {
      const step = 80 / size;

      for (let index = size; index > 0; index--) {
        output.push(hsl2hex({ ...hsl, l: step * index }));
      }

      break;
    }
    default: {
      const step = 360 / size;

      output.push(hsl2hex({ ...hsl, l: lightness ?? hsl.l, s: saturation ?? hsl.s }));

      for (let index = 1; index < size; index++) {
        const color = rotate(input, hsl.h + step * index, 'hex') as HEX;

        output.push(hsl2hex({ ...hex2hsl(color), l: lightness ?? hsl.l, s: saturation ?? hsl.s }));
      }

      break;
    }
  }

  return output.map(color => convert(color, format ?? colorFormat));
}
