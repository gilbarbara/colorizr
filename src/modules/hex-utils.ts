import { MESSAGES } from '~/modules/constants';
import { invariant } from '~/modules/invariant';
import { round } from '~/modules/utils';
import { isHex, isNumber } from '~/modules/validators';

import { Alpha } from '~/types';

/**
 * Add an alpha value to a hex string
 */
export function addAlphaToHex(input: string, alpha: Alpha): string {
  invariant(isHex(input), MESSAGES.inputHex);
  invariant(isNumber(alpha), MESSAGES.inputNumber);

  if (alpha >= 1) {
    return removeAlphaFromHex(input);
  }

  return `${removeAlphaFromHex(input)}${convertAlphaToHex(alpha)}`;
}

/**
 * Convert an alpha value to a hex value.
 */
export function convertAlphaToHex(input: Alpha): string {
  invariant(isNumber(input), MESSAGES.inputNumber);

  let alpha = input;

  if (input > 1) {
    alpha /= 100;
  }

  return Math.round(alpha * 255)
    .toString(16)
    .padStart(2, '0');
}

/**
 * Extract the alpha value from a hex string
 */
export function extractAlphaFromHex(input: string): number {
  invariant(isHex(input), MESSAGES.inputString);

  const alpha = input.substring(7, 9);

  if (!alpha) {
    return 1;
  }

  return round(parseInt(alpha, 16) / 255);
}

export function hexadecimalToNumber(input: string) {
  return round(parseInt(input, 16));
}

/**
 * Remove the alpha value from a hex string
 */
export function removeAlphaFromHex(input: string) {
  invariant(isHex(input), MESSAGES.inputHex);

  if (input.length === 5) {
    return input.substring(0, 4);
  }

  return input.substring(0, 7);
}
