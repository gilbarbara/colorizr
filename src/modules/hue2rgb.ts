import { invariant, isNumber, round } from './utils';

/**
 * Convert hue to RGB using chroma and median point
 */
export default function hue2rgb(point: number, chroma: number, h: number): number {
  invariant(isNumber(point) && isNumber(chroma) && isNumber(h), 'point, chroma and h are required');
  let hue = h;

  if (hue < 0) {
    hue += 1;
  }

  if (hue > 1) {
    hue -= 1;
  }

  if (hue < 1 / 6) {
    return round(point + (chroma - point) * 6 * hue, 4);
  }

  if (hue < 1 / 2) {
    return round(chroma, 4);
  }

  if (hue < 2 / 3) {
    return round(point + (chroma - point) * (2 / 3 - hue) * 6, 4);
  }

  return round(point, 4);
}
