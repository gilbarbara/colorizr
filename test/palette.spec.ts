import { messages } from 'modules/utils';
import palette from 'palette';

describe('palette', () => {
  it.each([
    ['#ff0044', undefined],
    ['#ff0044', { type: 'monochromatic' }],
    ['#ff0044', { lightness: 90, size: 10 }],
    ['#ff0044', { saturation: 70, size: 5 }],
    ['blue', undefined],
  ])('%p with %p should return the palette', (input, options) => {
    expect(palette(input, options)).toMatchSnapshot();
  });

  it('should fail with invalid parameters', () => {
    // @ts-ignore
    expect(() => palette([])).toThrow(messages.inputString);
    // @ts-ignore
    expect(() => palette('#f04', [])).toThrow('invalid options');
  });
});
