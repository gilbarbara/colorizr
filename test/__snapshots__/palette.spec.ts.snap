// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`palette > #ff0044 with undefined should return the palette 1`] = `
[
  "#ff0044",
  "#ff7700",
  "#88ff00",
  "#00ff77",
  "#0088ff",
  "#7700ff",
]
`;

exports[`palette > blue with undefined should return the palette 1`] = `
[
  "#0000ff",
  "#00ffff",
  "#0000ff",
  "#ff00ff",
  "#ff0000",
  "#ffff00",
]
`;

exports[`palette > hsl(136 100% 50%) with { type: 'monochromatic' } should return the palette 1`] = `
[
  "hsl(135.88 100% 80%)",
  "hsl(135.88 100% 66.67%)",
  "hsl(135.88 100% 53.33%)",
  "hsl(135.88 100% 40%)",
  "hsl(135.88 100% 26.67%)",
  "hsl(135.88 100% 13.33%)",
]
`;

exports[`palette > oklab(70.622% 0.1374 0.14283) with { size: 24 } should return the palette 1`] = `
[
  "oklab(70.622% 0.1374 0.14283)",
  "oklab(94.584% -0.10238 0.19437)",
  "oklab(90.666% -0.16276 0.18699)",
  "oklab(88.063% -0.20758 0.18213)",
  "oklab(86.838% -0.2302 0.17986)",
  "oklab(86.693% -0.23196 0.17514)",
  "oklab(87.073% -0.21846 0.14412)",
  "oklab(87.969% -0.19357 0.08364)",
  "oklab(89.461% -0.16467 0.00609)",
  "oklab(83.71% -0.12447 -0.07502)",
  "oklab(69.137% -0.07353 -0.15777)",
  "oklab(55.788% -0.03735 -0.24134)",
  "oklab(46.928% -0.03043 -0.29996)",
  "oklab(45.752% -0.0181 -0.30864)",
  "oklab(49.482% 0.062 -0.28823)",
  "oklab(56.342% 0.16055 -0.24879)",
  "oklab(65.012% 0.23986 -0.19849)",
  "oklab(68.552% 0.27621 -0.12937)",
  "oklab(65.697% 0.26965 -0.03907)",
  "oklab(63.811% 0.25 0.05045)",
  "oklab(62.933% 0.22951 0.11333)",
  "oklab(63.491% 0.21645 0.12736)",
  "oklab(68.37% 0.16101 0.13797)",
  "oklab(77.705% 0.07017 0.15804)",
]
`;

exports[`palette > oklch(47.642% 0.29956 274.93693) with { lightness: 90, size: 10 } should return the palette 1`] = `
[
  "oklch(87.316% 0.07112 297.48387)",
  "oklch(95.055% 0.04601 208.21388)",
  "oklch(88.969% 0.05328 268.93531)",
  "oklch(87.559% 0.07169 300.16324)",
  "oklch(90.238% 0.08548 324.36492)",
  "oklch(89.645% 0.06506 348.55536)",
  "oklch(90.26% 0.05068 28.66369)",
  "oklch(95.989% 0.05036 89.76108)",
  "oklch(97.473% 0.06852 122.65333)",
  "oklch(95.357% 0.08365 143.36983)",
]
`;

exports[`palette > rgb(255 230 109) with { saturation: 70, size: 5 } should return the palette 1`] = `
[
  "rgb(233 216 131)",
  "rgb(131 233 218)",
  "rgb(137 131 233)",
  "rgb(233 131 206)",
  "rgb(233 178 131)",
]
`;
